### PR TITLE
[Bounty] Adds the CVS recipient injector

### PR DESCRIPTION
**Pull request by BurgerLUA**
Link to bounty: https://forums.yogstation.net/threads/90-4-4-give-traitor-virologist-sentient-disease-kit.24327/

# Notice

WIP and untested. Don't give me money yet until I know it works and the payee is satisfied.

# TODO list

- [x] Test it.
- [x] Make sprites for it.
- [ ] Balance it.

# Document the changes in your pull request

Adds the CVS recipient injector, a 20TC Hijack only Virologist traitor item that silently injects someone with a disease and makes the mentioned disease sentient. Has 3 uses.

# Why is this good for the game.

image

# Icons

image

# Changelog

:cl:  
rscadd: Adds the CVS recipient injector, a 20TC Hijack only Virologist traitor item that silently injects someone with a disease and makes the mentioned disease sentient. Has 3 uses.
/:cl:

### CODE REVIEW

File: code/game/objects/items/sentient_disease_injector.dm [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	stored_vial = null
+
+	return TRUE
+
+/obj/item/sentient_disease_injector/attack(obj/item/I, mob/user, params)
+	return //Prevents damage.
+
+/obj/item/sentient_disease_injector/afterattack(atom/target, mob/user, proximity)
+
+	if(!proximity || !istype(target,/mob/living/carbon))
```
Comment (tattax) [inline code/game/objects/items/sentient_disease_injector.dm]:
```suggestion
	if(!proximity || !icarbon(target))
```

File: code/game/objects/items/sentient_disease_injector.dm [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+		return ..()
+
+	if(!used)
```
Comment (tattax) [inline code/game/objects/items/sentient_disease_injector.dm]: you mean (used)

File: code/game/objects/items/sentient_disease_injector.dm [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	used = TRUE
+
+	for(var/k in C.diseases) //Cure all existing diseases
+		var/datum/disease/D = k
+		D.cure(add_resistance = FALSE)
+
+	if(stored_vial && stored_vial.reagents.total_volume) //If there is a stored vial, inject.
+		var/list/injected = list()
+		for(var/k in stored_vial.reagents)
+			var/datum/reagent/R = k
```
Comment (tattax) [inline code/game/objects/items/sentient_disease_injector.dm]:
```suggestion
```

File: code/game/objects/items/sentient_disease_injector.dm [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+	used = TRUE
+
+	for(var/k in C.diseases) //Cure all existing diseases
+		var/datum/disease/D = k
+		D.cure(add_resistance = FALSE)
+
+	if(stored_vial && stored_vial.reagents.total_volume) //If there is a stored vial, inject.
+		var/list/injected = list()
+		for(var/k in stored_vial.reagents)
```
Comment (tattax) [inline code/game/objects/items/sentient_disease_injector.dm]:
```suggestion
		for(var/datum/reagent/R as anything in stored_vial.reagents)
```

File: code/game/objects/items/sentient_disease_injector.dm [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+			var/datum/reagent/R = k
+			injected += R.name
+		log_combat(user, C, "attempted to inject", src, "([english_list(injected)])")
+		stored_vial.reagents.reaction(C, INJECT, 1)
+		stored_vial.reagents.trans_to(C, stored_vial.reagents.total_volume, transfered_by = user)
+	else
+		log_combat(user, C, "attempted to inject", src)
+
+	if(length(C.diseases))
+		INVOKE_ASYNC(src, .proc/create_sentient_virus, target, user)
```
Comment (tattax) [inline code/game/objects/items/sentient_disease_injector.dm]:
```suggestion
		INVOKE_ASYNC(src, PROC_REF(create_sentient_virus), target, user)
```

File: code/game/objects/items/sentient_disease_injector.dm [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+		return FALSE //No candidates.
+
+	var/mob/dead/observer/selected = pick_n_take(candidates)
+	var/mob/camera/disease/virus = new/mob/camera/disease(get_turf(target))
+	virus.key = selected.key
+	message_admins("[ADMIN_LOOKUPFLW(virus)] has been made into a sentient disease by [ADMIN_LOOKUPFLW(usr)]'s [src].")
+	log_game("[key_name(virus)] was spawned as a sentient disease by [ADMIN_LOOKUPFLW(usr)]'s [src].")
+	//Mix and cure all existing diseases.
+	for(var/k in target.diseases)
+		var/datum/disease/D = k
```
Comment (tattax) [inline code/game/objects/items/sentient_disease_injector.dm]:
```suggestion
	for(var/datum/disease/D as anything in target.diseases)
```

Comment (tattax) [reply code/game/objects/items/sentient_disease_injector.dm]: could use better var names all around

File: code/game/objects/items/sentient_disease_injector.dm [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+		if(affecting.status != BODYPART_ORGANIC)
+			to_chat(user, span_warning("\The [src] won't work on a robotic limb!"))
+			return
+
+	to_chat(user, span_notice("You stealthily inject \the [C] with \the [src]."))
+
+	used = TRUE
+
+	for(var/k in C.diseases) //Cure all existing diseases
+		var/datum/disease/D = k
```
Comment (tattax) [inline code/game/objects/items/sentient_disease_injector.dm]:
```suggestion
	for(var/datum/disease/D as anything in C.diseases) //Cure all existing diseases
```

**Comment by tattax** [comment]
could you link the bounty pretty please

**Comment by BurgerLUA** [comment]
> could you link the bounty pretty please

Edited so that the bounty is linked in the OP.

File: code/game/objects/items/sentient_disease_injector.dm [hunk 8]
```diff
@@ -0,0 +1,126 @@
+/obj/item/sentient_disease_injector
+	name = "\improper CVS recipient injector"
+	desc = "It doesn't look like it prints recipts."
```
Comment (tattax) [inline code/game/objects/items/sentient_disease_injector.dm]:
```suggestion
	desc = "It doesn't look like it prints receipts."
```

**Comment by Rdrazga** [comment]
I struggle to understand the bounty this is for and why it would be better then just giving the virologist a traitor item that lets them break the normal 6 symptom limit without having to rely on a ghost who likely never played virology to not just instantly ruin the virus

**Comment by Rdrazga** [comment]
Or how insanely unbalanced this would be if said ghost did know virology due to having access to symptoms sentient viruses normally don’t get for balance reasons

**Comment by ehj236** [comment]
> Or how insanely unbalanced this would be if said ghost did know virology due to having access to symptoms sentient viruses normally don’t get for balance reasons

I could reverse it instead but I would like to expand upon Sentient disease so that future devs may take a look at it in the future as something that can eventually be fixed, besides viros with access to infinite symptoms by themselves would be literally gamebreaking in every regard, take note of APN(autophagocytosis necrosis) or ARDS(Acute respiratory distress syndrome) even Tumors while by themselves in 6 symptom restrictions. I would rather it be a two person job than someone who becomes a actual god with access to one of the most powerful systems in the entire game apart that rivals both xenobiology, nanites, robotics and even trumps the great works of genetics.

**Review by MajManatee — DISMISSED** [review]
Please don't have a sprite that's unused. Especially if its just to show what it looks like

**Comment by BurgerLUA** [comment]
> Please don't have a sprite that's unused. Especially if its just to show what it looks like

There are no unused sprites. The sprite (what it looks like) is not actually inside the .dmi file.

**Comment by BurgerLUA** [comment]
I have no way to test this since I have no friend, but it should work.

It compiles, so it works.

File: code/game/objects/items/sentient_disease_injector.dm (hunk 4 above)
Comment (BurgerLUA) [reply code/game/objects/items/sentient_disease_injector.dm]: This causes travis errors

**Review by cuackles — APPROVED** [review]
aproved

**Comment by BurgerLUA** [comment]
Okay so requesting a testmerge on this since I have no way to test it. Here are case scenarios of what would happen during a round with this in terms of shit breaking.

**Average Case Scenario**
The virus that the virologist made doesn't get properly merged with the sentient virus and you just have a sentient viruses injector with the sentient disease starting cough.

**Best Case Scenario**
It actually works flawlessly.

**Worst Case Scenario**
Runtime errors occur every time the sentient disease infects someone ~~or tries to open a menu~~ because it has symptoms that a sentient disease never actually is supposed to get.

**Comment by UltraBigRat** [comment]
why is this 20tc?

File: code/game/objects/items/sentient_disease_injector.dm [hunk 9]
```diff
@@ -0,0 +1,172 @@
+/obj/item/sentient_disease_injector
+	name = "\improper CVS recipient injector"
+	desc = "It doesn't look like it prints receipts..."
+
+	icon = 'yogstation/icons/obj/syringe.dmi'
```
Comment (tattax) [inline code/game/objects/items/sentient_disease_injector.dm:5]: if you're making the icon file modularized why not the dm file aswell?

File: code/game/objects/items/sentient_disease_injector.dm [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+	if(stored_vial)
+		if(stored_vial.reagents.total_volume > 0)
+			var/mutable_appearance/filling = mutable_appearance(icon,"[icon_state]_reagents")
+			filling.color = mix_color_from_reagents(stored_vial.reagents.reagent_list)
+			add_overlay(filling)
+		add_overlay(
+			image(
+				icon = icon,
+				icon_state = "[icon_state]_vial]"
```
Comment (ynot01) [inline code/game/objects/items/sentient_disease_injector.dm:41]: does this dangling ] need to be here?

File: code/game/objects/items/sentient_disease_injector.dm [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+	virus.key = selected.key
+	message_admins("[ADMIN_LOOKUPFLW(virus)] has been made into a sentient disease by [ADMIN_LOOKUPFLW(usr)]'s [src].")
+	log_game("[key_name(virus)] was spawned as a sentient disease by [ADMIN_LOOKUPFLW(usr)]'s [src].")
+	//Mix and cure all existing diseases.
+	for(var/datum/disease/D as anything in target.diseases)
+		if(D == virus)
+			continue
+		virus.disease_template.Mix(D)
+		D.cure(add_resistance = FALSE)
+
```
Comment (ynot01) [inline code/game/objects/items/sentient_disease_injector.dm:171]: Allow the sentient disease to pick a name via 	INVOKE_ASYNC(virus, TYPE_PROC_REF(/mob/camera/disease, pick_name))

**Review by ynot01 — CHANGES_REQUESTED** [review]
Runtime
```
[2023-07-23 23:50:34.801] [00:48:47] runtime error: Cannot modify null.appearance.
 - proc name: add infection (/mob/camera/disease/proc/add_infection)
 -   source file: disease_mob.dm,230
 -   usr: Alexander Crawford (/mob/living/carbon/human)
 -   src: Sentient Disease (/mob/camera/disease)
 -   usr.loc: the floor (86,151,2) (/turf/open/floor/plasteel)
 -   src.loc: the floor (75,155,2) (/turf/open/floor/plasteel)
 -   call stack:
 - Sentient Disease (/mob/camera/disease): add infection(Sentient Virus (/datum/disease/advance/sentient_disease))
 - Sentient Virus (/datum/disease/advance/sentient_disease): after add()
 - Sentient Virus (/datum/disease/advance/sentient_disease): infect(the monkey (796) (/mob/living/carbon/monkey), 0)
 - Sentient Virus (/datum/disease/advance/sentient_disease): infect(the monkey (796) (/mob/living/carbon/monkey), 0)
 - Sentient Virus (/datum/disease/advance/sentient_disease): try infect(the monkey (796) (/mob/living/carbon/monkey), 0)
 - the monkey (796) (/mob/living/carbon/monkey): ForceContractDisease(Sentient Virus (/datum/disease/advance/sentient_disease), 0, 1)
 - Sentient Disease (/mob/camera/disease): force infect(the monkey (796) (/mob/living/carbon/monkey))
 - the CVS recipient injector (/obj/item/sentient_disease_injector): create sentient virus(the monkey (796) (/mob/living/carbon/monkey), Alexander Crawford (/mob/living/carbon/human))
 - world: ImmediateInvokeAsync(the CVS recipient injector (/obj/item/sentient_disease_injector), /obj/item/sentient_disease_inj... (/obj/item/sentient_disease_injector/proc/create_sentient_virus), the monkey (796) (/mob/living/carbon/monkey), Alexander Crawford (/mob/living/carbon/human))
```

**Comment by Bibby0110** [comment]
Very stale.

**Pull request closed without merging**